crates/autopilot/src/database/onchain_order_events/ethflow_events.rs

```diff
@@ -0,0 +1,208 @@
+use crate::database::events::meta_to_event_index;
+use anyhow::{anyhow, Context, Result};
+use contracts::cowswap_onchain_orders::{
+    event_data::OrderPlacement as ContractOrderPlacement, Event as ContractEvent,
+};
+use database::{
+    ethflow_orders::EthOrderPlacement, events::EventIndex,
+    onchain_broadcasted_orders::OnchainOrderPlacement, orders::Order, PgTransaction,
+};
+use ethcontract::Event as EthContractEvent;
+use std::{collections::HashMap, convert::TryInto};
+use unzip_n::unzip_n;
+
+use super::{CustomOnchainOrderParsing, CustomParsedOnchaninData};
+
+unzip_n!(pub 3);
+
+pub struct EthFlowOnchainOrderParser {}
+
+#[async_trait::async_trait]
+impl<'a> CustomOnchainOrderParsing<'a, u32, EthOrderPlacement> for EthFlowOnchainOrderParser {
+    fn parse_custom_event_data(
+        &self,
+        contract_events: &[EthContractEvent<ContractEvent>],
+    ) -> Result<Vec<(EventIndex, CustomParsedOnchaninData<u32>)>> {
+        contract_events
+            .iter()
+            .filter_map(|EthContractEvent { data, meta }| {
+                let meta = match meta {
+                    Some(meta) => meta,
+                    None => return Some(Err(anyhow!("event without metadata"))),
+                };
+                let ContractEvent::OrderPlacement(event) = data;
+                let quote_and_valid_to_touple = convert_to_quote_id_and_user_valid_to(event);
+                match quote_and_valid_to_touple {
+                    Ok(touple) => Some(Ok((
+                        meta_to_event_index(meta),
+                        CustomParsedOnchaninData {
+                            quote_id: touple.0,
+                            additional_data: Some(touple.1),
```

Comment:
Would it work to have more structured additional data? For example, `EthFlowData {validTo: u32}`? I'd like to make more explicit what this field contains.